---
layout: default
title:  Course FAQs for PoreCamp 2016
---

# {{ page.title }}

Welcome to Penryn and the University of Exeter. We are delighted to welcome you to the
course and hope you will learn and enjoy much during the next week.

# Course structure

This is the second instalment of the PoreCamp series founded by Nick Loman and Josh
Quick. It was previously run at the University of Birmingham and will take a similar
structure with many of the same instructors. We will have an introduction to the
course and to each other on Monday 15th. Following this you will be assigned to a
rotation group which will spend a day on each of the following stations: QC & Library
prep, MinIon running and hacking, Bioinformatics analysis. Everyone will rotate
through each of these stations spending a day on each. The Friday of the course will
be devoted to bioinformatics analysis.

# What to bring

- Yourself
- A modern laptop with a reasonable screen size (e.g. 14-15 inch) and suitable power
cords and adaptors for UK power sockets (this will not be used for MinIon sequencing,
but rather to connect to resources to analyse MinIon data). Just in case, please also
bring an Ethernet cable and adaptor if required.
- Your favourite lab coat if you have one (we will be able to provide one if needed)
- Pens, markers and something suitable for use as a lab-book

# What will we be sequencing?

We aim to sequence 3-4 genomes from the 1000 fungal genomes project and publish the
results in collaboration with the project. We will be using the latest R9 flowcells
and preparing 2D barcoded libraries to run across approximately 3 flowcells each day.

# Can I bring my own material?

We are happy for participants to bring one sample of their own material for sequencing
provided it is either genomic or amplicon material. However, please do not rely on the
course to obtain results for your experiment. Flowcells are limited and it is unlikely
that you will generate sufficient data for your experiment.

If you do wish to bring your own material, please ensure you prepare no more than two
slides outlining what the sample is and why it is interesting. Where possible it should
also include including Agilent Tapestation or equivalent and estimates of
concentration. We will need a minimum of around 500ng-1ug of material. We will assess
the QC data as part of the course. A Tapestation will be onsite to perform QC.

Again, we reiterate - please do not rely on this course to generate data for your
experiment. We view the QC exercise as important to demonstrate the range of DNA
quality and its impact on library prep and sequencing rather than to ensure that we
obtain a good library and data from a given sample.

# New developments

There have been a few new developments from Oxford Nanopore over the past few weeks
including 1D rapid library prep kits and also the new Spot On flowcells. Whilst
weâ€™ll be focussing on the 2D preps, we should also have time for a demo of the 1D
prep. We will also be using the new Spot On flowcells.
