---
layout: default
title:  Course handbook for Porecamp 2016
---

# {{ page.title }}

# Group lectures and discussion

- History of second/third generation sequencing, nanopore in context
- Introduction to nanopore sequencing
- Introduction to applications (focus on microbial applications)
- Understanding nanopore data
- Nanopore workflow
- Group session: Designing a nanopore experiment

- Group dinner for participants and instructors.

# Library Prep

- Sample input considerations
- Fresh, frozen, FFPE
- Protocols (Native, PCR, PreCR, Low-input)
- Run through SQK-MAP-006 on different samples (as above)
- Josh run through and then students attempt their own

# MinION Running

- Configuring laptop correctly
- Using MinKNOW
- Loading the instrument
- Understanding MinKNOW
- Interpreting platform QC
- Interrogating runs in progress
- Discussion of read until - perhaps live version if I can.
- Metrichor
- Data management strategies
- Interrogation of ongoing runs: minoTour, poretools

# Bioinformatics

- Initial data handling
- Organisation of raw reads (poRe/Poretools)
- Extraction of FASTQ/A data (poRe/Poretools/minoTour)
- Gathering and visualization of run statistics (poRe/Poretools/minoTour!)
- Error correction and quality control (Nanocorr, NanoOK)
- Alignment of reads (Last, BWA, marginAlign)
- Event alignment (Nanopolish) - variant calling
- Scaffolding of short reads (SSPACE-LongRead)
- Hybrid de novo assembly (SPAdes)
- Discussion of pure de novo assembly (Celera, nanopolish)

